src/system/group.cr

```diff
@@ -0,0 +1,137 @@
+  def self.name?(gid : Int) : String?
+    check_gid_in_bounds(gid)
+    group_struct = LibC.getgrgid(gid)
+    return nil if !group_struct
+    String.new(group_struct.value.gr_name)
+  end
+
+  # Returns the group ID for the given group name.
+  # Raises `System::Group::NotFoundError` if no group exists.
+  def self.gid(group : String) : UInt32
+    gid?(group) || raise NotFoundError.new("Group #{group.inspect} was not found")
+  end
+
+  # Returns the group ID for the given group name.
+  # Returns `nil` if no group exists.
+  def self.gid?(group : String) : UInt32?
+    group.check_no_null_byte
+    group_struct = LibC.getgrnam(group)
+    return nil if !group_struct
+    group_struct.value.gr_gid
+  end
+
+  # Returns the group specified by the group ID.
+  # Raises `System::Group::NotFoundError` if not found.
+  def self.get(gid : Int) : Group
+    get?(gid) || raise NotFoundError.new("Group #{gid.inspect} was not found")
+  end
+
+  # Returns the user specified by the group ID.
+  # Returns `nil` if not found.
+  def self.get?(gid : Int) : Group?
+    check_gid_in_bounds(gid)
+    group_struct = LibC.getgrgid(gid)
+    return nil if !group_struct
+    new(group_struct.value)
+  end
+
+  # Returns the group specified by the groupname.
+  # Raises `System::Group::NotFoundError` if not found.
+  def self.get(groupname : String) : Group
+    get?(groupname) || raise NotFoundError.new("Group #{groupname.inspect} was not found")
+  end
+
+  # Returns the user specified by the username.
+  # Returns `nil` if not found.
+  def self.get?(groupname : String) : Group?
+    groupname.check_no_null_byte
+    group_struct = LibC.getgrnam(groupname)
+    return nil if !group_struct
+    new(group_struct.value)
+  end
+
+  # :nodoc:
+  private def initialize(group : LibC::Group)
+    @name = String.new(group.gr_name)
+    @gid = group.gr_gid
+    slice = Slice.new(group.gr_mem, limit: NGROUPS_MAX, read_only: true)
+    @member_names = Array.new(slice.size) { |idx| String.new(slice[idx]) }
+  end
+
+  # Returns the groups name as a `String`.
+  getter name : String
+
+  # Returns the group ID.
+  getter gid : UInt32
+
+  # Returns an `Array` of the groups members names as `String`s.
+  getter member_names : Array(String)
+
+  # Yields the groups members names as `String`s.
+  def each_member_name(&block : String -> Nil) : Nil
+    @member_names.each { |member| yield(member) }
+  end
+
+  # Returns an `Array` of the groups members as `User`s.
+  def members : Array(User)
+    @member_names.map { |member| User.get(member) }
+  end
+
+  # Yields the groups members as `User`s.
+  def each_member(&block : User -> Nil) : Nil
+    each_member_name { |member| yield(User.get(member)) }
+  end
+
+  # See `Object#hash(hasher)`
+  def hash(hasher)
+    @gid.hash(hasher)
+  end
+
+  # Compares this group with *other*, returning `-1`, `0` or `+1` depending if the
+  # group ID is less, equal or greater than the *other* group ID.
+  def <=>(other : Group) : Int
```

Comment:
I don't tyhink it makes sense to order by gid, is root really the lowest ordered user? no. If people want to order by gid they should be explciit about it.

You can replace this and `==` with the `def_hash_and_equals @gid` macro.